File: programs/programs/sea_invaders/src/instructions/tide.rs
//! `revive` ("the Tide"): a player who has just run out of
//! lives can pay to continue mid-level. The price rises one step on the
//! `Config.revive_ladder` each time it is used and ebbs back down by one
//! step for every full `cfg.ebb_seconds` that passes without a revive,
//! per the `effective`/`price` computation below.

use anchor_lang::prelude::*;

use crate::{
    errors::SeaError,
    instructions::ticket::{effective_tide, pay_split, BuyTicket},
    time,
};

#[event]
pub struct Revived {
    pub wallet: Pubkey,
    pub price: u64,
    pub tide: u8,
    pub tide_at: i64,
}

/// Reuses `BuyTicket`'s account set verbatim - the design doc says as much
/// ("accounts like `BuyTicket` (no catalog)") and this program already
/// reuses one `Accounts` struct across several instructions the same way
/// (`AdminOnly` backs `update_config`, `set_paused` and `set_test_clock`).
///
/// Takes `max_price`, mirroring `purchase(item_id, max_price)` in shop.rs:
/// the client quotes `cfg.revive_ladder[effective]` before sending the
/// transaction, and `max_price` caps what the chain may actually charge if
/// a concurrent revive (from another device) or an admin re-price raises
/// the ladder step between quote and execution.
pub fn revive(ctx: Context<BuyTicket>, max_price: u64) -> Result<()> {
    let cfg = &ctx.accounts.config;
    require!(!cfg.paused, SeaError::Paused);
    // Never call `Clock::get()` directly - `time::now` honours the
    // feature-gated test clock (see `src/time.rs`).
    let now = time::now(cfg)?;
    let week = time::week_of(time::day_of(now));
    require!(ctx.accounts.week_pool.week == week, SeaError::WrongWeekPool);

    let tide = ctx.accounts.player.tide;
    let tide_at = ctx.accounts.player.tide_at;
    let effective = effective_tide(tide, tide_at, now, cfg.ebb_seconds)?;
    let price = cfg.revive_ladder[effective as usize];
    require!(price <= max_price, SeaError::PriceChanged);

    pay_split(
        &ctx.accounts.token_program,
        &ctx.accounts.skr_mint,
        &ctx.accounts.wallet_token,
        &ctx.accounts.wallet,
        &ctx.accounts.vault,
        &ctx.accounts.treasury,
        price,
        cfg.purchase_pool_bps,
    )?;

    let p = &mut ctx.accounts.player;
    p.tide = effective.saturating_add(1).min(7);
    p.tide_at = now;

    emit!(Revived {
        wallet: p.wallet,
        price,
        tide: p.tide,
        tide_at: p.tide_at,
    });
    Ok(())
}

File: backend/src/services/tide.js
// The Tide (design doc §1/§2, `revive`): a rising/ebbing SKR price for continuing mid-level after
// the last life is lost. `effectiveTide` mirrors `programs/.../instructions/ticket.rs::effective_tide`
// exactly (same test cases, ported), so the quote a player sees is the price the program will
// actually charge at the same clock - `revive_ladder` and `tide`/`tide_at` always come from the
// chain, never hardcoded here.
import { getConfig, getConfirmedInstructions, getPlayer, getTokenBalance } from '../chain/readers.js';
import { buildReviveTx } from '../chain/txs.js';
import { hasRevive } from '../chain/verify.js';
import * as loadoutDb from '../db/loadout.js';
import { dayOf, weekOf } from './dailySeed.js';
import { ShopError } from './shop.js';

/**
 * The tide step actually in effect `now`, after ebbing back from `tide` by one step per full
 * `ebbSeconds` since `tideAt` - `tideAt === 0` (never revived) always yields 0 elapsed steps.
 * Integer arithmetic only, truncating toward zero exactly like the program's `i64` division.
 */
export function effectiveTide(tide, tideAt, now, ebbSeconds) {
  const elapsedRaw = tideAt === 0 ? 0 : Math.trunc((now - tideAt) / ebbSeconds);
  const elapsed = Math.max(0, Math.min(tide, elapsedRaw));
  return tide - elapsed;
}

/** `{ tide, effective, priceSkr, nextStep: { priceSkr, inSeconds } | null }` for `player` (`null` = no `Player` PDA yet) against `config`, at `now`. Shared by `quoteRevive` and `issueRevive` so both read the chain exactly once. */
function quoteFromPlayer(player, config, now) {
  const tide = player ? player.tide : 0;
  const tideAt = player ? player.tideAt : 0;
  const ladder = config.reviveLadder;
  const ebbSeconds = config.ebbSeconds;

  const effective = effectiveTide(tide, tideAt, now, ebbSeconds);
  const priceSkr = Number(ladder[effective]) / 1e6;

  // `tide - effective` is the elapsed-steps count `effectiveTide` actually applied (already clamped
  // to `[0, tide]`), so the next boundary is exactly one more full `ebbSeconds` after `tideAt`. Only
  // meaningful once a revive has actually happened (`tideAt !== 0`) - `tide > 0, tideAt === 0` cannot
  // occur on chain (a revive always stamps `tide_at = now`), but a quote for it must still say the
  // price never ebbs, not report a boundary computed from a `tideAt` of 0.
  let nextStep = null;
  if (effective > 0 && tideAt !== 0) {
    const elapsedSteps = tide - effective;
    const boundary = tideAt + (elapsedSteps + 1) * ebbSeconds;
    nextStep = { priceSkr: Number(ladder[effective - 1]) / 1e6, inSeconds: boundary - now };
  }

  // The whole ladder in SKR, so the app can show its range (design: `Price ladder 25 -> 120 SKR`) without hardcoding prices.
  const ladderSkr = ladder.map((v) => Number(v) / 1e6);

  return { tide, effective, priceSkr, nextStep, ladderSkr };
}

export async function quoteRevive({ wallet, now }) {
  const [player, config] = await Promise.all([getPlayer(wallet), getConfig()]);
  return quoteFromPlayer(player, config, now);
}

/**
 * Builds the unsigned `revive` transaction at the currently quoted price - `create_player` first
 * when the wallet has no `Player` PDA yet (`revive` reuses `buy_ticket`'s accounts, which require
 * one; see `issueTicket` in `services/records.js`). Throws `not_enough_skr` (with `needSkr`/`haveSkr`)
 * if the wallet's SKR balance is short.
 */
export async function issueRevive({ wallet, now }) {
  const [player, balance, config] = await Promise.all([getPlayer(wallet), getTokenBalance(wallet), getConfig()]);
  const quote = quoteFromPlayer(player, config, now);
  const priceBaseUnits = config.reviveLadder[quote.effective];
  if (balance < priceBaseUnits) {
    throw new ShopError('not_enough_skr', 'Not enough SKR to revive', { needSkr: quote.priceSkr, haveSkr: Number(balance) / 1e6 });
  }

  const week = weekOf(dayOf(now));
  const createsPlayer = !player;
  const envelope = await buildReviveTx(wallet, { week, treasury: config.treasury, createPlayer: createsPlayer });
  return { ...envelope, priceSkr: quote.priceSkr, createsPlayer };
}

/**
 * Confirms a submitted `revive` transaction: verifies it actually invokes our program's `revive`
 * instruction, signed by `wallet` (global-constraints.md's Phase 3B addition), before refreshing
 * the loadout cache. Re-callable while the transaction is not yet visible (`confirmed: false`).
 */
export async function confirmRevive({ wallet, signature }) {
  const parsed = await getConfirmedInstructions(signature);
  if (parsed.status === 'missing') return { confirmed: false };
  if (parsed.status === 'failed') throw new ShopError('revive_failed', 'The revive transaction failed on chain');
  if (!hasRevive(parsed.instructions, { wallet })) {
    throw new ShopError('invalid_transaction', 'Transaction does not match the revive request');
  }

  const player = await getPlayer(wallet);
  const tide = player ? player.tide : 0;
  const tideAt = player ? player.tideAt : 0;
  await loadoutDb.upsertLoadout(wallet, { tide, tideAt, inventory: player ? player.inventory : 0n });
  return { confirmed: true, ok: true, tide, tideAt };
}
